src/sema/expression/expression_impl.rs

```diff
@@ -0,0 +1,1203 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::sema::ast::{ArrayLength, Expression, Namespace, RetrieveType, Type};
+use crate::sema::contracts::is_base;
+use crate::sema::diagnostics::Diagnostics;
+use crate::sema::eval::{eval_const_number, eval_const_rational};
+use crate::sema::expression::compatible_mutability;
+use num_bigint::{BigInt, Sign};
+use num_rational::BigRational;
+use num_traits::ToPrimitive;
+use num_traits::{FromPrimitive, Zero};
+use solang_parser::diagnostics::Diagnostic;
+use solang_parser::pt;
+use solang_parser::pt::CodeLocation;
+use std::cmp::Ordering;
+
+impl Expression {
```

Comment:
It feels weird to me having that in not in the same module where the `Expression` enum is defined (however, the `ast` module would blow up in size if we do this). How about putting this inside `mod.rs` instead?

Reply:
I concur, this also means we get rid of the filename `expression_impl.rs` which is a bit strange. Names should describe what things do, not what syntax they use. 

This reminds me of C++ pimpl which surely an abomination :see_no_evil:

Reply:
Moving `impl Expression` to `mod.rs` will enlarge the file to more than three thousand lines. I really would like to avoid that. Another option is to create a new file for `fn expression` and leave the `impl Expression` in `mod.rs`. Is this better?

I agree with @xermicus that `impl Expression` should be in the same file as `enum Expression`, but I think @seanyoung is not keen on the idea to split the AST definition in multiple files.

Reply:
The AST is already defined in multiple files: yul is defined elsewhere.

Also, we have `impl Type` in various places, certainly not just in `ast.rs`. I don't think this is necessarily bad, but then again I don't think large files are so bad either.

Reply:
> Another option is to create a new file for `fn expression` and leave the `impl Expression` in `mod.rs`. Is this better?

What about this?